### DISCUSSION TITLE
A tip: use asyncElegantOta, a lifesaver

### DISCUSSION THREAD

**Discussion by nousian** [hatlabs/discussions#60] [Ideas]
I suggest usage of asyncElegantOta is added to the example code provided.

With just few extra lines it really makes life onboard boat so much easier as you don't need to disconnect any wires to update the software (requires wifi naturally):

```
#include <AsyncTCP.h>
#include <ESPAsyncWebServer.h>
#include <AsyncElegantOTA.h>

// ---

void serveHttp(AsyncWebServerRequest *request)  {
    request->send(200, "text/plain", "Hi! I am ESP32.");
}

// ---

void setupOTA() {  
  server.on("/", HTTP_GET, serveHttp);

  AsyncElegantOTA.begin(&server);    // Start ElegantOTA
  server.begin();
  Serial.println("HTTP server started");  
}

void setup {
  setupOTA();
  // do other stuff maybe as well
}
```

**Comment by nousian** [comment]
Also, if one uses the included button, its handy to have a code to simulate button presses remotely. I used async udp for that, my button shifts displayed page on the device itself:

```
void buttonPressed() {
  if (debug) Serial.write("buttonPressed\r\n");

  if ((millis() - lastDebounceTime) > debounceDelay) {
    buttonPressCount++;
    displayPage = buttonPressCount % 2;
    lastDebounceTime = millis();
  }
}

void udpRXCallBack(AsyncUDPPacket &packet) {
  if (debug) Serial.write("udpRXCallBack\r\n");

  udp_rx_count++;
  
  std::string str( packet.data(), packet.data() + packet.length() );

  if (str.rfind("button", 0) == 0) {
    buttonPressed();
  }
}

```

Something along those lines.

Btw, I am willing to contribute my complete project code someplace, if there is interest. It does NMEA0183 and NMEA2000 wirelessly to both directions and bunch of other stuff as well. It is quite heavily customized to my own use case, but I think someone could find useful pieces to take from there.

I use it to connect to Signalk and Opencpn, with no usb connected. NMEA2000 is from internal canbus, NMEA0183 is optocoupler input/output with softserial. 

NMEA2000 is sent over UDP encapsulated in NMEA0183 $PCDINs, which Signalk reads out of the box, NMEA0183 as raw UDP packets.

**Comment by hwater** [comment]
Hello Nousian, i would like to test your NMEA ESP32 Code. I like the cconcept to do wireless UDP communication.

pls send it to holgerw @ me.com

keep on - Holger

**Comment by mairas** [comment]
Unfortunately, licensing of that library makes it a non-starter for too many use cases. But perhaps as importantly, I am steering away from any libraries depending in ESPAsyncTCP. As far as I have understood, its TCP stack implementation is fundamentally broken and it will potentially cause subtle issues that are difficult to trace.

See this video for more information: https://www.youtube.com/watch?v=O2W8uyC-Gio

**Reply by JohnySeven** [reply]
@mairas in TWatchSK firmware I made fully fuctional websocket and http handling for Signal K connection, if you need some inspiration or help moving SensESP to ESP-IDF networking let me know.
